handle CTRL-C on foreground node

phillyphil91 opened the pull request:
<!-- Thank you for sending a pull request :heart: -->

## Current behavior
Foreground nodes are not properly terminated on `CRTL-C`. They remain in `ockam node list`, even after `CTRL-C`.

<!-- Please describe the current behavior of the code before the changes in this pull request is applied. -->

## Proposed changes
A listener for `CTRL-C` is introduced and handles gracefully shutting down the foreground node including removing the node state.
Addresses Issue: #4372

<!-- Please describe the changes proposed in this pull request. -->
<!-- If this pull request resolves an already recorded bug or a feature request, please add a link to that issue. -->

## Checks

<!-- To help us review and merge this pull request quickly, please confirm the following:  -->

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/edit/main/CONTRIBUTORS.csv) file in the github web UI and create a PR, this will mark the commit as verified.

<!-- Looking forward to merging your contribution!! -->

phillyphil91 commented:
Hi.
A couple of notes/questions from my side.

Instead of using the `ctrlc` crate, I found that `tokio` already provides similar functionality with `tokio::signal::ctrl_c` if I understood the workings of both solutions correctly. Both allow for a `CTRL-C` handler to be created. `ctrlc` apparently does this via a dedicated thread. So maybe using the `tokio` version is more close to what we want and part of the dependencies already anyways.

I hope I understood the issue correct that `CTRL-C` previously caused the node to remain in the `ockam node list` ouput even though it was run as a foreground node and thus should not? Now if you start a foreground node and do `CTRL-C` is disappears and leaves no trace in `ockam node list`. This was not the case before. Of course this only handles `CTRC-C` which I think is `SIGINT` at least on unix system. So it probably doesn't handle `SIGTERM` correctly in general.

@adrianbenavides  you mentioned to call the `NodeState::delete()` function after catching `CTRL-C`. I believe `opts.state.nodes.delete(&node_name, false)?` does this.

Regarding `let _ = tokio::spawn(stop_foreground_node(ctx, opts, node_name)).await?;` I was not sure whether `stop_foreground_node()` really needs to be passed to `tokio::spawn()`. On the other hand if I just `await` `stop_foreground_node()` directly or even move the function body directly into `run_foreground_node()` then this would block until `CTRL-C` is caught and thus never reach the `OK(())` part right? Not sure if that's actually what happens and if that's what we want. 
Thanks for help on this. I know it's not much of a great change yet.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -351,6 +351,20 @@ async fn run_foreground_node(
         stop_node_on_eof(&mut ctx, &opts, &node_name).await?;
     }
 
+    let _ = tokio::spawn(stop_foreground_node(ctx, opts, node_name)).await?;
```
Whether you use tokio or ctrlc to handle the signals, they can be handled in a separate thread and run in the background... or handled in the main thread blocking its execution. What approach to use will depend on the use case.

With tokio, if you `await` the spawned task, you will stop the execution there until the inner future completes. So, you just want to spawn the thread and continue: `tokio::spawn(stop_foreground_node(ctx, opts, node_name));`. In this case, you can get away without communicating back to the main thread because you can stop the context from wherever you want as long as you are in an async runtime.

With `ctrlc`, the handler is a simple closure (non-async), so you have to communicate back to the main thread where you will be able to stop the context. A side effect of this is that you should run the `stop_node_on_eof` in a background tokio task and set up the ctrlc signal in the main thread, similar to how it's shown in the [docs example](https://docs.rs/ctrlc/latest/ctrlc/#example). The reason of this is explained on the other comment.

And yeah, this issue doesn't require too many lines, but it's full of hidden details and traps 😆

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -351,6 +351,20 @@ async fn run_foreground_node(
         stop_node_on_eof(&mut ctx, &opts, &node_name).await?;
     }
```
This `stop_node_on_eof` function runs a loop and blocks the execution until it receives an EOF character, so the signal handler should be registered before this block.

We have to support these two loops at the same time and exit on the first one that happens, so we can only block the main thread with only one of the loops.

adrianbenavides commented:
First off, thank you so much for taking a look into this!


> Instead of using the `ctrlc` crate, I found that `tokio` already provides similar functionality with `tokio::signal::ctrl_c`

Yeah, they are equivalent but `ctrlc` support SIGINT, SIGTERM and SIGHUP by default. I think it w

> I hope I understood the issue correct that `CTRL-C` previously caused the node to remain in the `ockam node list` ouput even though it was run as a foreground node and thus should not? Now if you start a foreground node and do `CTRL-C` is disappears and leaves no trace in `ockam node list`. This was not the case before. Of course this only handles `CTRC-C` which I think is `SIGINT` at least on unix system. So it probably doesn't handle `SIGTERM` correctly in general.

It might worked on your example but the signal handler is not set up correctly. I'll leave a comment explaining how to use it properly.

> @adrianbenavides you mentioned to call the `NodeState::delete()` function after catching `CTRL-C`. I believe `opts.state.nodes.delete(&node_name, false)?` does this.

Actually, now that I think of it, I think the correct thing to do is to stop the node, rather than delete it. You can do that with:

```rust
ctx.stop().await?;
opts.state.nodes.get(node_name)?.kill_process(false)?;
```

> Regarding `let _ = tokio::spawn(stop_foreground_node(ctx, opts, node_name)).await?;` I was not sure whether `stop_foreground_node()` really needs to be passed to `tokio::spawn()`. On the other hand if I just `await` `stop_foreground_node()` directly or even move the function body directly into `run_foreground_node()` then this would block until `CTRL-C` is caught and thus never reach the `OK(())` part right? Not sure if that's actually what happens and if that's what we want. Thanks for help on this. I know it's not much of a great change yet.

I'll leave a comment about this.

adrianbenavides commented:
Finally, one way to test this would be to run a foreground node with the `--exit-on-eof` argument so you set up the two handlers. Then you can try stopping the node with either of the signals supported and verify that the node is stopped correctly.

phillyphil91 commented:
Hi @adrianbenavides,

once again, thanks for the helpful comments. I thought I had it, but then realized I didn't :smile: 

So the `ctrlc` crate is more useful for us here because it handles `SIGTERM`, `SIGINT`, etc. But other than that has no advantage over `tokio::signal::ctrl_c`?

I tried to implement everything according to the `ctrlc` example which includes running a loop that checks if `signal_terminate` is set to `true` and then run the node stop/ kill process.

This seems to work when only `--foreground` is set. Once `--foreground` and `--exit-on-eof` are set this does not work anymore. The node is still stopped, but you have to manually confirm `CTRL-C`. Sending `SIGTERM` also stops the node but the process is still running etc.

I figured since the monitoring of `EOF` and `CTRL-C`, etc. requires looping it would make sense to spawn both them in a new `tokio task`. And then either could run the `ctx.stop()` and `kill_process()` functions and then everything would terminate. 
But this doesn't seem to be the case and I am not sure why. It seems like either one of the two or the main thread is still running even after one of them has successfully executed. Maybe it's neither of the two tasks can actually kill the entire process if the other one is still running (possibly in another thread)?

Tried just sending one of to a new spawned tasks and await the other at the end, but this didn't seem to really help either.

phillyphil91 commented:
Maybe the solution is to use `tokio::task::spwan_blocking` or `std::threat::spawn` because both functions are blocking indeed.

With `tokio::task::spwan_blocking` `CTRL-C` at least seems to work as expected, but `EOF` does not.

I have to take another look.

phillyphil91 commented:
Okay so I think the current version works, even though I am not entirely sure how.

`stop_foreground_node()` now is in wrapped in `spawn_blocking` and `stop_node_on_eof()` now just blocks the main thread at the end. 

I used the approach with the `Arc` because `ctx` doesn't implement `Clone`. If it did I could have just cloned the arguments to `stop_foreground_node()` in the closure I believe.

adrianbenavides commented:
I haven't gone through the PR in detail but given how complex is this getting I'd recommend using an mpsc channel so you can send the `stop node` signal back to the main thread from background tasks (i.e. the ctrlc and the eof functions). The main advantage of this approach is that you won't have to pass the `Context` to the spawned threads, and therefore, you won't have to deal with the side effects of cloning the context.

Let me know if you need more context on how to do that and I'll provide a detailed example.

phillyphil91 commented:
> I haven't gone through the PR in detail but given how complex is this getting I'd recommend using an mpsc channel so you can send the `stop node` signal back to the main thread from background tasks (i.e. the ctrlc and the eof functions). The main advantage of this approach is that you won't have to pass the `Context` to the spawned threads, and therefore, you won't have to deal with the side effects of cloning the context.
> 
> Let me know if you need more context on how to do that and I'll provide a detailed example.

Thanks. I will try to solve this using an mpsc channel. I will get back to you if I have any questions.

The current version is not suitable yet?

adrianbenavides commented:
> The current version is not suitable yet?

It's a valid solution, but I think it's worth exploring other simpler alternatives. The code in the `ockam_command` crate is usually the first contact with the ockam repository for most of the people that contributes to it for the first time. It's also common that those contributors have never written any rust code before. If I put myself in that situation, I know for sure this code would blow my mind 😆

phillyphil91 commented:
> > The current version is not suitable yet?
> 
> It's a valid solution, but I think it's worth exploring other simpler alternatives. The code in the `ockam_command` crate is usually the first contact with the ockam repository for most of the people that contributes to it for the first time. It's also common that those contributors have never written any rust code before. If I put myself in that situation, I know for sure this code would blow my mind :laughing: 

That makes sense. Plus to be fair I am not even sure how this one works but sending both to `tokio::task::spawn_blocking` does not. :confused: 
I will have a look at the mpsc channel alternative.

phillyphil91 commented:
Hi @adrianbenavides 

Over the weekend I tried to implement this in an `async` fashion in every possible way under the sun, I believe :sweat_smile: 
But any combination with `tokio::spawn` or `tokio::task::spawn_blocking` did not work and always left the program hanging. Even when receiving the appropriate message on the channel or reaching the final `Ok(())`.

It is working with dedicated OS threads as in the current version. I am not sure if that's okay giving the environments that Ockam is supposed to run in. I still used `tokio::sync::mpsc::channel()` to not block the `run_foreground_node` function.

I removed the `stop_node_on_eof` function and integrated that straight into `run_foreground_node` in a slightly different fashion. Let me know what you think.

P.s. I still have to clean up the other files that I changed for a previous approach

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -347,32 +346,34 @@ async fn run_foreground_node(
+    })
+    .expect("Error setting Ctrl+C handler");
+
+    // Spawn a thread to monitor STDIN for EOF
+    let tx_clone = tx.clone();
+    std::thread::spawn(move || {
+        let mut buf = String::new();
+        match io::stdin().read_to_string(&mut buf) {
+            Ok(_) => tx_clone.blocking_send(1).unwrap(),
+            Err(e) => eprintln!("Error reading STDIN: {}", e),
+        };
+    });
```
This should only be executed `if cmd.exit_on_eof`. Also, the logic is wrong as it will exit upon any read line.

adrianbenavides replied:
Here is an example of how it could look:

```rust
let mut buffer = Vec::new();
let mut handle = io::stdin().lock();
handle
    .read_to_end(&mut buffer)
    .expect("Error reading from stdin");
info!("! EOF received");
let _ = tx_clone.blocking_send(());
```

adrianbenavides replied:
If it does what it has to do, pressing `ctrl+d` should trigger the EOF handler.

phillyphil91 replied:
I can change that to your version, but I thought since `read_to_string()` says that it 

```
Read(s) all bytes until EOF in this source, appending them to buf.

If successful, this function returns the number of bytes which were read and appended to buf.
```

I figured that it would jut continue to read `STDIN` until it receives `EOF` from somewhere and not exit on any line?
Or is not calling `.lock()` the issue here?

adrianbenavides replied:
mm good point. To be honest I'm not sure of the differences between the two methods for our use-case. You can keep your version and try if it works as expected. If it reacts to `ctrl+d` then we are good.

phillyphil91 replied:
I think mine was working too, but I changed it to yours because  the function name `read_to_end()` is more closely related to what we actually want to do :smile:

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -347,32 +346,34 @@ async fn run_foreground_node(
+    // Register a handler for the Ctrl+C signal
+    let tx_clone = tx.clone();
+    ctrlc::set_handler(move || {
+        tx_clone.blocking_send(1).unwrap();
```
If you want to return a dummy value you can return the unit `()` value.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -347,32 +346,34 @@ async fn run_foreground_node(
-            Err(_) => unreachable!(),
+    if cmd.foreground || cmd.exit_on_eof {
+        // Shutdown on SIGINT, SIGTERM, SIGHUP or EOF
+        if let Some(_) = rx.recv().await {
```
The `rx.recv().await` must be always checked. In the default case, you will register the ctrlc signal and you want to wait for that. So, the `if cmd.foreground || cmd.exit_on_eof` should not be there.

phillyphil91 replied:
I thought the default case was to not watch either: `CTRL-C `and `EOF`. but only do so when `--foreground` is set (`CTRL-C`) and of course when `--exit-on-eof` is set (`EOF`).

So my idea was to run both transmission channels and then run `recv()` only when either `--foreground` or `--exit-on-eof` is set. 
But I am just realizing that then `EOF` would also always be monitored even if just `--foreground` is set. Which is wrong.

If `ctrlc` should always be handled, I will change it to that.

adrianbenavides replied:
Note that we are already in the `run_foreground_node` function, so we can assume `--foreground` is passed. Does that make sense?

phillyphil91 replied:
of course it does. I totally overlooked the context of the function. sorry, that makes total sense

adrianbenavides replied:
no worries, I also had to double-check this one 😆

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -347,32 +346,34 @@ async fn run_foreground_node(
+        // Shutdown on SIGINT, SIGTERM, SIGHUP or EOF
+        if let Some(_) = rx.recv().await {
+            ctx.stop().await?;
+            opts.state.nodes.get(&node_name)?.kill_process(false)?;
```
I'd switch these two lines. Stopping the context should be the last operation to run before exiting an ockam runtime.

adrianbenavides reviewed (commented):
I've been playing around a little with this and I've managed to make it work with the changes I've described below. I can push my changes if you get stuck, but I think you are pretty close now! 💪

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -347,32 +346,36 @@ async fn run_foreground_node(
+    // Register a handler for SIGINT, SIGTERM, SIGHUP
+    let tx_clone = tx.clone();
+    ctrlc::set_handler(move || {
+        tx_clone.blocking_send(()).unwrap();
```
Add message here:

```rust
opts.shell.write_line(format!(
    "{} Ctrl+C signal received",
    "!".light_yellow(),
    &cmd.node_name
))?;
```

phillyphil91 replied:
Had to modify this a bit, but works now, hopefully as you would expect. I wasn't sure if we want the node_name printed as well. let me know.

Also I just unwraped the `shell.write_line()`. Is this okay? Or rather `expect()` with a more meaningful message?

adrianbenavides replied:
>  I wasn't sure if we want the node_name printed as well. let me know.

Sorry, copy-paste error. No need to show the node_name. I added a separate comment about the unwrapped `write_line()`.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -347,32 +346,36 @@ async fn run_foreground_node(
+            handle
+                .read_to_end(&mut buffer)
+                .expect("Error reading from stdin");
+            info!("EOF received!");
```
Change this message to:

```rust
opts.shell.write_line(format!(
    "{} EOF received",
    "!".light_yellow(),
    &cmd.node_name
))?;
```

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -347,32 +346,36 @@ async fn run_foreground_node(
+    // Shutdown on SIGINT, SIGTERM, SIGHUP or EOF
+    if let Some(_) = rx.recv().await {
+        opts.state.nodes.get(&node_name)?.kill_process(false)?;
+        ctx.stop().await?;
```
Add message:

```rust
opts.shell.write_line(format!(
    "{}Node stopped successfully",
    "✔︎".light_green(),
    &cmd.node_name
))?;
```

adrianbenavides reviewed (commented):
Looks like it's working now perfectly! There are a couple of minor things before we can merge it:
- Rebase on top of latest develop
- Print some messages to inform the user of what's going on. Once you have the changes from develop, you will be able to print the messages as pointed out in the comments

Once you are done with that, you will need to:
- squash all commits
- write a valid commit message, for example: `feat(rust): handle ctrlc signal on foreground nodes`

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -352,8 +353,46 @@ async fn run_foreground_node(
+        opts.state.nodes.get(&node_name)?.kill_process(false)?;
+        ctx.stop().await?;
+        opts.shell
+            .write_line(format!("{} Node stopped successfully", "✔︎".light_green()).as_str())
```
The "✔︎" includes a trailing space at the end, so it should be `{}Node`.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -352,8 +353,46 @@ async fn run_foreground_node(
+                .shell
+                .write_line(format!("{} EOF received", "!".light_yellow()).as_str())
+                .unwrap();
+            let _ = tx_clone.blocking_send(());
```
Let's send the message first and then log the message. In this context, we can ignore the error as it can't be recovered and we don't care if we can't write to stderr:

```rust
let _ = tx_clone.blocking_send(());
let _ = opts_clone.shell.write(...);
```

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -352,8 +353,46 @@ async fn run_foreground_node(
+    let mut opts_clone = opts.clone();
+    ctrlc::set_handler(move || {
+        tx_clone.blocking_send(()).unwrap();
+        opts_clone
+            .shell
+            .write_line(format!("{} Ctrl+C signal received", "!".light_yellow()).as_str())
+            .unwrap();
```
Same here about ignoring the error.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/create.rs:
```diff
@@ -352,8 +353,46 @@ async fn run_foreground_node(
+    }
+
+    // Shutdown on SIGINT, SIGTERM, SIGHUP or EOF
+    if let Some(_) = rx.recv().await {
```
There is a clippy warning left to be fixed here.

phillyphil91 committed:
feat(rust): handle ctrlc signal on foreground nodes

adrianbenavides reviewed (approved):
Ready to merge, thanks @phillyphil91 for another great contribution! 💪

phillyphil91 commented:
thank you @adrianbenavides for your help again.
